media-source-respec.html

```diff
@@ -2719,6 +2711,280 @@ <h3>Event Summary</h3>
+        is an implementation defined condition where the user agent determines that
+        it currently has enough data to play the presentation without stalling for a
+        meaningful period of time. This condition is constantly evaluated to determine
+        when to transition the value of {{ManagedMediaSource/streaming}}. These transitions
+        indicate when the user agent believes it has enough data buffered or it
+        needs more data respectively.
+      </p>
+      <p>
+        Being <dfn>able to retrieve and buffer data in an efficient way</dfn>
+        is an implementation defined condition where the user agent determines that
+        it can fetch new data in an energy efficient manner while able to achieve the desired memory usage.
+      </p>
+      <ol class="algorithm">
+        <li>Run the {{MediaSource}} [=MediaSource/SourceBuffer Monitoring=] algorithm.</li>
+        <li>Let |can play uninterupted and efficiently| be a flag that is true if the {{HTMLMediaElement/buffered}} attribute contains a {{TimeRanges}} that includes the current playback position and [=enough managed data to ensure uninterrupted playback=]
+          and is [=able to retrieve and buffer data in an efficient way=]
+          <dl class="switch">
+            <dt>If |can play uninterupted and efficiently| is not equal to {{ManagedMediaSource/streaming}}, [=queue an element task=] on the [=media element=]
+              that runs the following steps:
+            </dt>
+            <dd>
+              <ol>
+                <li>Set |this| {{ManagedMediaSource/streaming}} attribute to |can play uninterupted and efficiently|.</li>
+                <li>If |can play uninterupted and efficiently| is false [=fire an event=] called {{startstreaming}} at the {{ManagedMediaSource}}.</li>
+                <li>Else [=fire an event=] called {{endstreaming}} at the {{ManagedMediaSource}}.</li>
+              </ol>
+            </dd>
+          </dl>
+        </li>
+      </ol>
+
+      <h4><dfn data-dfn-for="ManagedMediaSource">Memory Cleanup</dfn></h4>
+      <ol>
+        <li>
+          <dl class="switch">
+            <dt>
+              For each |buffer:ManagedSourceBuffer| in [=this=]'s
+              {{MediaSource/sourceBuffers}}:
+            </dt>
+            <dd>
+              <ol>
+                <li>Run the |buffer|'s [=ManagedSourceBuffer/memory cleanup=] algorithm.</li>
+              </ol>
+            </dd>
+          </dl>
+        </li>
+      </ol>
+    </section>
+
+    <section data-dfn-for="BufferedChangeEvent">
+      <h2><dfn>BufferedChangeEvent</dfn> interface</h2>
+      <pre class="idl">
+        [Exposed=(Window,DedicatedWorker)]
+        interface BufferedChangeEvent : Event {
+          constructor(DOMString type, BufferedChangeEventInit eventInitDict);
+
+          [SameObject] readonly attribute TimeRanges addedRanges;
+          [SameObject] readonly attribute TimeRanges removedRanges;
```

Comment:
It's not obvious how an app might differentiate bufferedchangeevent's removedRanges from being caused by memory cleanup versus explicit remove that happened near in time to the memory cleanup (or other reasons like re-opening the ManagedMediaSource after it previously had been in "ended" readyState.) Is such differentiation desired by MMSE apps? If such differentiation were needed later, would this API shape be flexible to provide that information to the app?

Reply:
There is no intention to separate the two, and I don't believe players would care. In fact I believe they should not differentiate: particularly because when appending data, it was possible for data to be evicted first (through the prepare append algorithm). The idea that the player could somehow assume that when they appended data nothing else was removed was bad, and has been the source of a lot of error in the wild.

How they would deal with content added and/or removed would pretty much be independent, then to check that there's data where they expect it to be.

Reply:
Further, making the distinction would reveal implementation details that can't be relied on, as they can change from one browser version to another or on different hardware. 

So, I don't think we say anything further here.